### DISCUSSION TITLE
Make GuiBase a subclass of QMainWindow

### DISCUSSION THREAD

**Discussion by Neverhorst** [Ulm-IQO/qudi-core#15] [Ideas]
This idea has been roaming through my heads' backlog for quite some time....

## Current State
Currently the `GuiBase` class is a `QObject` subclass like logic and hardware. 
References to the widgets and especially the `QMainWindow` to show are solely at the mercy of the user implementation. The only structure the `GuiBase` class imposes, is the abstract `show()` method that must be provided by every GUI module implementation.

It has become a de-facto standard in _qudi_ to store the "QMainWindow" instance of the GUI module in `self._mw` on activation and any `show()` implementation looks the same. 
Thus any GUI module currently looks like this:
```python
MyGuiModule(GuiBase):
    ...
    def on_activate(self) -> None:
        self._mw = QMainWindow()
        ...

    def show(self) -> None:
        self._mw.show()
        self._mw.activateWindow()
        self._mw.raise_()
```

## The Problem
Besides being a minor case of code duplication, the current structure does not allow for the `GuiBase` class (or any other object) to manage and automize certain generic tasks related to window management, e.g.:
- Show/Hide window
- GUI module deactivation upon closing the main window 
- Window geometry save/restore
- Generic `QMainWindow` event handling
- Standardized window title management
- Generic multi-monitor window handling

This is because there is no generic access to the `QMainWindow` instance provided by the GUI module.  Since every _qudi_ GUI module does instantiate a `QMainWindow` to show (otherwise it would be a logic module), this is a completely unnecessary inconvenience.

## Proposal
I mainly see two ways to tackle this problem:

1. Include an abstract main window instance property/getter in `GuiBase` so that the GUI module implementation must provide access to the `QMainWindow` instance it uses as main window. 

2. Make `GuiBase` a direct subclass of `QMainWindow`.

The advantage of __option 1__ is less "namespace pollution" of the GUI module with a ton of `QMainWindow` attributes. The disadvantage is a non-obvious requirement for people implementing GUI modules that they need to handle manually.  It is also less strict regarding typing and not enforcing GUI modules to have a `QMainWindow` always.

The disadvantage of __option 2__ is "namespace pollution" of the GUI module. The advantages are strict enforcement of a main window for every GUI module (like it should be, otherwise it's a logic module), clear object oriented approach that is widely considered good practice and less things to take care of for _qudi_ module programmers.

I would strongly suggest option 2. A GUI module in _qudi_ only needs to implement `on_activate` and `on_deactivate` like logic and hardware and could then look like e.g.:
```python
from PySide2.QtWidgets import QPushButton, QSpinBox, QLabel, QGridLayout, QWidget
from qudi.core.module import GuiBase

class MyMainWindow(GuiBase):
    """ Layout of the QMainWindow for MyGuiModule. 
    If the main window is defined by a .ui file, import it here instead of the __init__ below.
    """
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)
        self.calc_button = QPushButton('Calculate')
        self.left_operand_spinbox = QSpinBox()
        self.right_operand_spinbox = QSpinBox()
        self.result_label = QLabel(' = ?')
        
        layout = QGridLayout()
        layout.addWidget(self.left_operand_spinbox, 0, 0)
        layout.addWidget(self.QLabel(' + '), 0, 1)
        layout.addWidget(self.right_operand_spinbox, 0, 2)
        layout.addWidget(self.result_label, 0, 3)
        layout.addWidget(self.calc_button, 1, 0, 1, 4)

        widget = QWidget()
        widget.setLayout(layout)
        self.setCentralWidget(widget)


class MyGuiModule(MyMainWindow):
    """ The "actual" qudi GUI module
    """
    def on_activate(self) -> None:
        self.calc_button.clicked.connect(self._calculate_result)

    def on_deactivate(self) -> None:
        self.calc_button.clicked.disconnect()

    def _calculate_result(self) -> None:
        result = self.left_operand_spinbox.value() + self.right_operand_spinbox.value()
        self.result_label.setText(f' = {result:d}')
```

I would like to implement that very soon before the "official" release of `qudi-core`.
If you have any input on the matter or different ideas, please feel free to post/discuss them here.

**Comment by timoML** [comment]
After looking at the QMainWindow documentation, I think a qudi window has a pretty clear "is a QMainWindows" relationship. Thus, I feel **option 2** is a fair way to go.
